pkgs/development/libraries/catboost/default.nix

```diff
@@ -0,0 +1,135 @@
+{ lib
+, config
+, stdenv
+, fetchFromGitHub
+, autoPatchelfHook
+, cmake
+, libiconv
+, llvmPackages
+, ninja
+, openssl
+, python3Packages
+, ragel
+, yasm
+, zlib
+, cudaSupport ? config.cudaSupport or false
+, cudaPackages ? {}
+, pythonSupport ? false
+, symlinkJoin
+}:
+let
+  inherit (cudaPackages) cudaVersion;
+
+  cuda-common-redist = with cudaPackages; [
+    cuda_cccl
+    libcublas
+  ];
+  cuda-native-redist = symlinkJoin {
+    name = "cuda-native-redist-${cudaVersion}";
+    paths = with cudaPackages; [
+      cuda_cudart
+      cuda_nvcc
+    ] ++ cuda-common-redist;
+  };
+
+  cuda-redist = symlinkJoin {
+    name = "cuda-redist-${cudaVersion}";
+    paths = cuda-common-redist;
+  };
+in
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "catboost";
+  version = "1.2.1";
+
+  src = fetchFromGitHub {
+    owner = "catboost";
+    repo = "catboost";
+    rev = "refs/tags/v${finalAttrs.version}";
+    hash = "sha256-JcbGEkBQe4En/1g1BY8ytV7cWN9CjujMrThIrJ0o7dU=";
+  };
+
+  patches = [
+    ./remove-conan.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace cmake/common.cmake \
+      --replace  "\''${RAGEL_BIN}" "${ragel}/bin/ragel" \
+      --replace "\''${YASM_BIN}" "${yasm}/bin/yasm"
+
+    shopt -s globstar
+    for cmakelists in **/CMakeLists.*; do
+      sed -i "s/OpenSSL::OpenSSL/OpenSSL::SSL/g" $cmakelists
+      ${lib.optionalString (lib.versionOlder cudaVersion "11.8") ''
+        sed -i 's/-gencode=arch=compute_89,code=sm_89//g' $cmakelists
+        sed -i 's/-gencode=arch=compute_90,code=sm_90//g' $cmakelists
```

Comment:
Is this dropping the support for 89, 90? Is there a way to specify the list of gencodes for which to build the kernels?

Reply:
Seeing how moving `cuda_*` to `*[bB]uildInputs` has worked, and how the build logs seem to mention FindCUDAToolkit.cmake: https://gist.github.com/SomeoneSerge/395dfc734cc220aae928cd9d501acabd#file-gistfile0-txt-L52

...I think it should be possible to control the target cuda architectures by setting https://cmake.org/cmake/help/latest/variable/CMAKE_CUDA_ARCHITECTURES.html, unless upstream overrides the variable. The configurable way to choose capabilities at the moment would be to use `cudaPackages.cudaFlags`, which respect `config.cudaCapabilities` (and whatever they might get replaced with later). Examples:
https://github.com/NixOS/nixpkgs/blob/b5213d5b0f6bda0fe098ff98890599ff63dfdd5f/pkgs/development/libraries/science/math/faiss/default.nix#L103
https://github.com/NixOS/nixpkgs/blob/b5213d5b0f6bda0fe098ff98890599ff63dfdd5f/pkgs/development/libraries/science/math/magma/generic.nix#L78-L81

Reply:
I just tested this in https://github.com/SomeoneSerge/nixpkgs/pull/8/commits/9780f86c4a3938c166959895cad8cdb2b3104fde and it doesn't work because upstream passes some ad hoc flags instead of using `CMAKE_CUDA_ARCHITECTURES`:

https://github.com/catboost/catboost/blob/ad3934317eb5357781e2389f1c24582c01f738ca/catboost/cuda/ctrs/CMakeLists.linux-x86_64-cuda.txt#L34-L56 https://github.com/catboost/catboost/blob/ad3934317eb5357781e2389f1c24582c01f738ca/cmake/cuda.cmake#L163-L165

I think we should contact upstream about changing that